### ISSUE TITLE
Begin Outputting `AnnData` file format

### ISSUE THREAD

**Issue by ilan-gold** [COMBINE-lab/alevin-fry#151]
I'm not sure if here or in [`pyroe`](https://pyroe.readthedocs.io/en/latest/) this issue belongs, but I think (at least from my limited understanding), it would be cool if the sparse matrix output of `alevin` could be `AnnData` on disk.  It seems like https://docs.rs/anndata/latest/anndata/ would be a good candidate for the `i/o`.

It seems like `infer` and `quant` would be the two candidates for them.  This issue is mostly about opening a discussion about this feature, not being prescriptive or anything!  I'm new to this ecosystem :)

**Comment by rob-p** [comment]
Hi @ilan-gold,

  Indeed, this would be great! This would be trivial to do in pyroe, as that is in Python and has immediate access to all relevant scVerse packages.

  That said, it would be cool to avoid a conversion program at all and be able to output AnnData directly (or auto convert to it if desired from within Rust). Do you know if there is an AnnData crate for rust? Otherwise, I know there’s and HDF5 crate, though that I think that pulls in the (kinda flakey?) C-based dependency.

—Rob

**Comment by kaizhang** [comment]
@ilan-gold @rob-p Please see here for an example of how to use anndata Rust package to output a cellxgene count matrix: https://github.com/regulatory-genomics/precellar/blob/a8d2fca438baf596a2be7cc6432364c7ef3534fe/precellar/src/transcript/quantification.rs#L80. Note the latest version use zstd compression under the hood. When you open the h5ad file in python, you may need to import hdf5plugin.

**Comment by ilan-gold** [comment]
@kaizhang Thanks.  I guess with `zarr` the issue is that `zarrs` is not feature complete with all of the python codecs for string handling (although I think it is v2 compliant).  Is there an option for uncompressed hdf5? I think we use that regularly without issue.

**Comment by ilan-gold** [comment]
Hmmm @rob-p This could be an edge case that `anndata-rs` is not handling.  If you send the file I can have a look to try to see.

**Comment by ilan-gold** [comment]
Ok so here are some (maybe informative) code snippets:

```python
import anndata as ad
import numpy as np

ad.AnnData(X=None, layers={"foo": np.arange(4).reshape((2, 2)), "bar": np.arange(4).reshape((2, 2))}) # errors, probably erroneously
```

```python
import anndata as ad
import numpy as np

adata = ad.AnnData(X=np.arange(4).reshape((2, 2)), layers={"foo": np.arange(4).reshape((2, 2)), "bar": np.arange(4).reshape((2, 2))})
del adata.X
adata.write_h5ad("foo.h5ad")
ad.read_h5ad("foo.h5ad").X == None
```
so if you're having trouble reading the data back in, it's tough to say without the file.  So I stand by my original point, please send.

**Comment by rob-p** [comment]
Hi @ilan-gold,

  So [here's my quick and dirty test](https://github.com/rob-p/test_anndata) to convert a USA matrix market format output from `alevin-fry` into an AnnData object.  Right now, I get around the above issue by just setting the `X` layer to be an empty matrix so that it takes essentially no space.  Obviously this code needs to be made much cleaner and pulled out into it's own function / crate, but it can be used to create files to test for downstream compatibility --- currently I can load up what it produces in python using `anndata.read`, so that's a good start!

**Comment by ilan-gold** [comment]
Nice, the goal here would be to do this directly from the in-memory representation in alevin-fry? Or would the intermediate `mtx` format be needed even here "in production?"

**Comment by rob-p** [comment]
I think both paths would be possible.  My thought on the migration path and testing it out would be the following.

Phase 1:
  * build a new crate (current proposed name `af-anndata`) that has the improved version of this conversion function.
  * pull this crate into `simpleaf`, and have `simpleaf` take a flag (or just alter the default behavior) to convert the `alevin-fry` output to the `anndata` format after quantification is complete.  This requires no changes to the current `alevin-fry`, but removes the user's need to perform an extra conversion step manually as long as they are using `simpleaf`. We could optionally choose to retain or remove the original CSR representation, though for compatibility with other downstream ecosystems, we'd want to make sure that any relevant packages are also updated to accept the anndata input (e.g. fishpond cc @DongzeHE, @mikelove).
  
Phase 2:
  * If phase 1 goes well, and we iron out any usability or performance wrinkles, then we could move the `af-anndata` dependency from `simpleaf` to `alevin-fry` itself, and output the `anndata` representation from `alevin-fry` directly, completely sidestepping the conversion.  We'd probably want to deprecate and remove the CSR output at the same time to avoid having to support and maintain 2 separate output formats.
  * I don't think we're currently dealing with individual datasets this large, but one thing worth considering is how challenging (if at all) it would be to do chunked output in the `anndata` format.  Specifically, one of the benefits of CSR is that we consider each cell barcode as it's own row, and CSR is amenable to row-by-row construction and output.  This means we never need to hold the whole output matrix in memory and instead only have to hold a number of cells proportional to the number of threads the user is using for quantification.  Such a scheme would also be useful if it could be done with the `anndata` object as well.

These are my initial thoughts, but I'm open to discussion or feedback on any of these points!

**Comment by ilan-gold** [comment]
> I don't think we're currently dealing with individual datasets this large, but one thing worth considering is how challenging (if at all) it would be to do chunked output in the anndata format. Specifically, one of the benefits of CSR is that we consider each cell barcode as it's own row, and CSR is amenable to row-by-row construction and output. This means we never need to hold the whole output matrix in memory and instead only have to hold a number of cells proportional to the number of threads the user is using for quantification. Such a scheme would also be useful if it could be done with the anndata object as well.

This would require custom code + handling via zarr/hdf5 but I think should be doable (if not so performant on zarr).  But that's a long way off, and that's just a first impression.

Otherwise I like the plan.  Seems like a good way to gradually introduce the functionality.

**Comment by rob-p** [comment]
> Otherwise I like the plan. Seems like a good way to gradually introduce the functionality.

Awesome!  Then I'll get started on this (i.e. grab the `af-anndata` crate name, move the test code to the corresponding repository and begin improving and robustifying the implementation).  I'll tag you in that repo once it's up with some basic code, and when the first version of that is complete, I'll start integrating the relevant functionality into `simpleaf`, which should be trivial once the `af-anndata` crate is ready!

**Comment by mikelove** [comment]
> any relevant packages are also updated to accept the anndata input

This should be fine as long as it's covered by a CRAN package e.g.:

https://cran.r-project.org/web/packages/anndata/readme/README.html

But GitHub only packages wouldn't work as dependencies.

**Comment by ilan-gold** [comment]
> optionally choose to retain or remove the original CSR representation, though for compatibility with other downstream ecosystems, we'd want to make sure that any relevant packages are also updated to accept the anndata input

I definitely think putting this behind a flag is a good way to start, but yea, these `AnnData` objects should be extremely simple, at least at first.  So I would guess there are minimal changes needed for compat downstream.

**Comment by rob-p** [comment]
Hi @ilan-gold,

  I've started implementing this in a library. That repo is [af-anndata].  In the mean time, if you want to take a look at a converted output, you can find one I made and uploaded [here](https://umd.box.com/s/broiy8ysq6xqkrgb170bfk5mlpab356b).  This has layers for the spliced, unspliced and ambiguous counts (and corresponding `varm` entries for the gene symbol IDs as well). Feel free to poke around and:
  
  (1) let me know if you have any suggestion on how to avoid the "empty `X` layer" hack.
  (2) let me know if you have any other suggestions on output format, naming in the object, etc.
  
We can continue discussion here or, if you'd prefer, we can pick it up in the `af-anndata` repo!

**Comment by ilan-gold** [comment]
Hi @rob-p I'd be still interested in what caused https://github.com/COMBINE-lab/alevin-fry/issues/151#issuecomment-2579079579.  You shouldn't have to make an empty CSR matrix so it would be good to understand why the file could not be read back in (maybe the same issue I raised in AnnData: https://github.com/scverse/anndata/issues/1816, even likely).  Or you could wait for me to fix that issue and then try it out later.  If we need to go back and re-do, then that's what it is.

**Comment by rob-p** [comment]
The feature this issue asked for exists now, so I'm closing it — with one caveat below.

`simpleaf quant --anndata-out` writes `quants.h5ad` alongside the usual matrix-market output, and `simpleaf multiplex-quant` does the same. The conversion lives in a separate crate, [`af-anndata`](https://github.com/COMBINE-lab/af-anndata), which reads the quantification output and writes AnnData directly — so, as you suggested, it goes through the Rust `anndata` crate rather than a Python round trip. Run metadata is embedded in `uns`.

@ilan-gold — your follow-up question is still unanswered, and closing this issue shouldn't make it disappear. You asked why an empty CSR matrix had to be constructed for the file to be readable, and whether it was the same problem as scverse/anndata#1816. That's a question about `af-anndata`'s implementation rather than about whether alevin-fry emits AnnData, and it will get better attention as its own issue on the [af-anndata](https://github.com/COMBINE-lab/af-anndata) tracker than buried at the bottom of a closed feature request here. Please do open one (or say the word and we will), and thank you for the review comments along the way — they improved the output.
